backend/dal/Services/Admin/Concrete/AgencyService.cs

```diff
@@ -100,35 +100,133 @@ public Agency Get(int id)
+            }
+        }
+
         /// <summary>
         /// Updates the specified agency in the datasource.
         /// </summary>
-        /// <param name="entity"></param>
-        /// <exception type="KeyNotFoundException">Entity does not exist in the datasource.</exception>
+        /// <param name="agency"></param>
+        /// <exception type="KeyNotFoundException">agency does not exist in the datasource.</exception>
         /// <returns></returns>
-        public override void Update(Agency entity)
+        public override void Update(Agency agency)
         {
-            entity.ThrowIfNull(nameof(entity));
+            agency.ThrowIfNull(nameof(agency));
+
+            var original = this.Context.Agencies.Find(agency.Id) ?? throw new KeyNotFoundException();
+            var updatedUsers = new List<User>();
+
+            // If the agency has become a sub-agency, or a parent-agency then users will need to be updated.
+            if (original.ParentId.HasValue && !agency.ParentId.HasValue)
+            {
+                // This agency has become a parent agency, all users associated with it will be removed
+                // because users should only be associated with a single parent.
+                updatedUsers.AddRange(this.Context.Users.Where(u => u.Agencies.Any(a => a.AgencyId == original.ParentId)));
+
+                // Remove any user who are linked to the agency.
+                var users = this.Context.Users.Include(u => u.Agencies).Where(u => u.Agencies.Any(ua => ua.AgencyId == agency.Id));
+                users.ForEach(u => u.Agencies.Clear());
+                this.Context.UpdateRange(users);
+            }
+            else if (!original.ParentId.HasValue && agency.ParentId.HasValue)
+            {
+                // This agency has become a sub-agency, all original users need their list of agencies reduced to only this agency.
+                updatedUsers.AddRange(this.Context.Users.Include(u => u.Agencies).Where(u => u.Agencies.Any(a => a.AgencyId == agency.ParentId)));
+            }
+            else if (original.ParentId != agency.ParentId)
+            {
+                // Remove the sub-agency from curretly linked users and add it to the users who belong to the new parent agency.
+                updatedUsers.AddRange(this.Context.Users.Where(u => u.Agencies.Any(a => a.AgencyId == original.ParentId)));
+                updatedUsers.AddRange(this.Context.Users.Where(u => u.Agencies.Any(a => a.AgencyId == agency.ParentId)));
+            }
 
-            var agency = this.Context.Agencies.Find(entity.Id) ?? throw new KeyNotFoundException();
+            if (original.IsDisabled != agency.IsDisabled)
+            {
+                if (agency.ParentId.HasValue && agency.IsDisabled)
+                {
+                    // Remove the sub-agency from users.
+                    updatedUsers.AddRange(this.Context.Users.Where(u => u.Agencies.Any(a => a.AgencyId == agency.ParentId)));
```

Comment:
Same answer as above.

Reply:
So if I disable Fraser Health, then this will add all of the users who belong to the MOH (because MOH is my parent agency) to this updatedUsers list. All of the users in that list then get the disabled Fraser Health agency added to them (but not in the DB) with this code:

```
updatedUsers.Distinct().ForEach(u =>
            {
                this.Context.Entry(u).State = EntityState.Detached;
                agency.Users.Add(new UserAgency(u, agency));
            });
``` 
and then that list of users gets sent to keycloak right? 

Presumably for all of the users in my agency that just removes the agency from their keycloak user.

How does this show up in keycloak for a user in another MOH subagency? I guess for them this is just a no-op?

For users that belong to the parent agency, I guess we aren't worried because our Logic based inheritence should prevent them from using the disabled agency?

Reply:
If you disable a sub-agency, all users who belong to it or the parent-agency will be returned by this function.  The Controller will then update Keycloak for all users in the list.

Users who belong to the parent agency will be updated to no longer have access to the disabled agency.
Users who belong only to the sub-agency will no longer have access to data.

Keycloak keeps an attribute array of all agencies the user has access to (i.e. if the user is linked to a parent agency `[1, 5, 20, 30]`, or if the user is linked only to a sub-agency `[5]`).

There are still issues when an agency is disabled for a user currently logged in.  This is because their lookup list of agencies still contains the disabled agency.  If they refresh it will be gone.  Additionally their token will contain the agency that was disabled until it is refreshed.